tools/make_patch.c

```diff
@@ -0,0 +1,476 @@
+	buffer->item_size = item_size;
+	buffer->size = 0;
+	buffer->capacity = 0x10;
+	buffer->data = xmalloc(buffer->capacity * item_size);
+	return buffer;
+}
+
+void append_to_buffer(struct Buffer *buffer, const void *item) {
+	if (buffer->size >= buffer->capacity) {
+		buffer->capacity = (buffer->capacity + 1) * 2;
+		buffer->data = xrealloc(buffer->data, buffer->capacity * buffer->item_size);
+	}
+	memcpy(buffer->data + (buffer->size++ * buffer->item_size), item, buffer->item_size);
+}
+
+void free_buffer(struct Buffer *buffer) {
+	free(buffer->data);
+	free(buffer);
+}
+
+int get_address_type_limit(int address) {
+	return address >= 0x8000 && address < 0xa000 ? 0xa000 // VRAM
+	     : address >= 0xa000 && address < 0xc000 ? 0xc000 // SRAM
+	     : address >= 0xc000 && address < 0xf000 ? 0xd000 // WRAM
+	     : address >= 0xff80 && address < 0xffff ? 0x10000 // HRAM
+	     : 0x4000; // ROM
+}
+
+void append_symbol(struct Symbol **symbols, const char *name, int bank, int address) {
+	struct Symbol *symbol = xmalloc(sizeof(*symbol) + strlen(name) + 1);
+	symbol->address = address;
+	symbol->offset = bank > 0 ? address + (bank - 1) * get_address_type_limit(address) : address;
+	strcpy(symbol->name, name);
+	symbol->next = *symbols;
+	*symbols = symbol;
+}
+
+void free_symbols(struct Symbol *list) {
+	for (struct Symbol *next; list; list = next) {
+		next = list->next;
+		free(list);
+	}
+}
+
+const struct Symbol *find_symbol(const struct Symbol *symbols, const char *name) {
+	size_t namelen = strlen(name);
+	for (const struct Symbol *symbol = symbols; symbol; symbol = symbol->next) {
+		size_t symbolnamelen = strlen(symbol->name);
+		if (namelen > symbolnamelen) {
+			continue;
+		}
+		const char *symbolname = symbol->name;
+		if (name[0] == '.') {
+			// If `name` is a local label, compare it to the local part of `symbol->name`
+			symbolname += symbolnamelen - namelen;
+		}
+		if (!strcmp(symbolname, name)) {
+			return symbol;
+		}
+	}
+	error_exit("Error: Unknown symbol: %s\n", name);
+	return NULL;
+}
+
+const struct Symbol *find_symbol_cat(const struct Symbol *symbols, const char *prefix, const char *suffix) {
+	char *symname = xmalloc(strlen(prefix) + strlen(suffix) + 1);
+	strcpy(symname, prefix);
+	strcat(symname, suffix);
+	const struct Symbol *symbol = find_symbol(symbols, symname);
+	free(symname);
+	return symbol;
+}
+
+void parse_symbol_value(char *input, int *bank, int *address) {
+	char *buffer = xmalloc(strlen(input) + 1);
+	strcpy(buffer, input);
+
+	char *buffer2 = strchr(buffer, ':');
+	if (buffer2) {
+		// Parse symbol's bank:address
+		*buffer2++ = '\0';
+
+		char *endptr_bank, *endptr_address;
+		*bank = strtol(buffer, &endptr_bank, 16);
+		if (*bank > 0xff || *bank < 0) {
+			*bank = -1;
+		}
+		*address = strtol(buffer2, &endptr_address, 16);
+		if (*address > 0xffff || *address < 0) {
+			*address = -1;
+		}
+
+		if (endptr_bank != buffer + strlen(buffer) || endptr_address != buffer2 + strlen(buffer2)) {
```

Comment:
Instead of calculating `strlen`, the recommended method of checking `strtol`'s return value is `endptr != str && *endptr == '\0'` (check that at least one digit was read, and check that the character after the last digit is, in this case, the end of the string).

Reply:
So `endptr == str || *endptr` would check for failure.